src/moi_interop.jl

```diff
@@ -79,18 +79,162 @@ function update!(moi_f::MOI.VectorAffineFunction, fs::Vector{<:AffineFunction},
+canonical_function_type(::Type{<:LinearTerm}, ::Type{T}) where {T} = AffineFunction{T}
+canonical_function_type(::Type{<:AffineFunction}, ::Type{T}) where {T} = AffineFunction{T}
+canonical_function_type(::Type{<:Vector{<:AffineFunction}}, ::Type{T}) where {T} = Vector{AffineFunction{T}}
+canonical_function_type(::Type{<:QuadraticTerm}, ::Type{T}) where {T} = QuadraticFunction{T}
+canonical_function_type(::Type{<:QuadraticFunction}, ::Type{T}) where {T} = QuadraticFunction{T}
+
+
+# moi_to_native_type
+moi_to_native_type(::Type{MOI.ScalarAffineFunction}) = AffineFunction{T} where T
+moi_to_native_type(::Type{MOI.ScalarQuadraticFunction}) = QuadraticFunction{T} where T
+moi_to_native_type(::Type{MOI.VectorAffineFunction}) = Vector{AffineFunction{T}} where T
+moi_to_native_type(::Type{MOI.SingleVariable}) = Nothing
+
+
+# Objective
+struct Objective{E, F}
+    expr::WrappedExpression{E}
+    f::F
+end
+
+function Objective(::Type{T}, expr) where T
+    val = evalarg(expr)
+    E = canonical_function_type(typeof(val), T)
+    converted = @expression convert(E, expr)
+    wrapped = convert(WrappedExpression{E}, converted)
+    f = make_moi_equivalent(E)
+    F = typeof(f)
+    update!(f, wrapped())
+    Objective{E, F}(wrapped, f)
+end
+
+function update!(objective::Objective{E, F}, optimizer::MOI.AbstractOptimizer, varmap) where {E, F}
+    update!(objective.f, objective.expr(), varmap)
+    MOI.set!(optimizer, MOI.ObjectiveFunction{F}(), objective.f)
+    nothing
+end
+
+
+# Constraint
+mutable struct Constraint{E, F<:MOI.AbstractFunction, S<:MOI.AbstractSet}
+    expr::WrappedExpression{E}
+    f::F
+    set::S
+    modelindex::MOI.ConstraintIndex{F, S}
+    optimizerindex::MOI.ConstraintIndex{F, S}
+
+    function Constraint(::Type{T}, expr, set::S) where {T, S<:MOI.AbstractSet}
+        val = evalarg(expr)
+        E = canonical_function_type(typeof(val), T)
+        converted = @expression convert(E, expr)
+        wrapped = convert(WrappedExpression{E}, converted)
+        f = make_moi_equivalent(E)
+        F = typeof(f)
+        update!(f, wrapped())
+        new{E, F, S}(wrapped, f, set)
+    end
+
+    function Constraint(f::F, set::S) where {F<:MOI.AbstractFunction, S<:MOI.AbstractSet}
+        E = Nothing
+        wrapped = convert(WrappedExpression{E}, @expression nothing)
+        new{E, F, S}(wrapped, f, set)
+    end
 end
 
-function MOI.ScalarQuadraticFunction(f::QuadraticFunction{T}) where T
-    ret = MOI.ScalarQuadraticFunction(MOI.ScalarAffineTerm{T}[], MOI.ScalarQuadraticTerm{T}[], f.affine.constant[])
-    update!(ret, f)
+function update!(constraint::Constraint, optimizer::MOI.AbstractOptimizer, varmap)
+    update!(constraint.f, constraint.expr(), varmap)
+    MOI.set!(optimizer, MOI.ConstraintFunction(), constraint.optimizerindex, constraint.f)
+    nothing
 end
+update!(constraint::Constraint{Nothing}, optimizer::MOI.AbstractOptimizer, varmap) = nothing
+
+
+# Constraints
```

Comment:
This is essentially doing the same thing as https://github.com/JuliaOpt/MathOptInterface.jl/blob/575647cfa344ed7656e0eee16537017184a011a8/src/Utilities/model.jl#L448 right? It would be great to eventually extract out the common behavior so we don't have to keep re-implementing this sort of thing.

Reply:
Well, kind of, in that both use metaprogramming to create a type and functions, but that's pretty generic. Not sure if there's anything common that can be extracted.